### PR TITLE
FILE_SUFFIX fixes

### PR DESCRIPTION
**Pull request by 57300**
* Fix doubling `FILE_SUFFIX` on `boards/*.overlay files`
* Use default `FILE_SUFFIX` in `zephyr_file()` (wasn't working in sysbuild)
* Bonus: use `FILE_SUFFIX` to simplify a sample's configuration

### CODE REVIEW

**Review by nordicjm — COMMENTED** [review]
Seems it is not working in sysbuild, share/sysbuild/cmake/modules/sysbuild_kconfig.cmake
```
if(DEFINED SB_CONF_FILE)
  # SB_CONF_FILE already set so nothing to do.
elseif(DEFINED ENV{SB_CONF_FILE})
  set(SB_CONF_FILE $ENV{SB_CONF_FILE})
elseif(EXISTS      ${APP_DIR}/sysbuild.conf)
  set(SB_CONF_FILE ${APP_DIR}/sysbuild.conf)
else()
  # Because SYSBuild is opt-in feature, then it is permitted to not have a
  # SYSBuild dedicated configuration file.
endif()
```
would need to be updated to support it

**Comment by nordicjm** [comment]
Just noticed another issue, `-DSB_FILE_SUFFIX` does not work for `<app_dir>/sysbuild/<app>/prj_<suffix>.conf`

**Comment by 57300** [comment]
> Seems it is not working in sysbuild, share/sysbuild/cmake/modules/sysbuild_kconfig.cmake

Right, this wasn't implemented for `SB_CONF_FILE`. Added another commit.

> Just noticed another issue, `-DSB_FILE_SUFFIX` does not work for `<app_dir>/sysbuild/<app>/prj_<suffix>.conf`

There is no `SB_FILE_SUFFIX`. The regular `FILE_SUFFIX` should work here.

**Comment by nordicjm** [comment]
> > Maybe I'm mixing it up with sysbuild. If you use `-Dsysbuild_FILE_SUFFIX=x` can you check it works?
> 
> `-D<image>_FILE_SUFFIX` works. `sysbuild` is not a valid prefix afaik.

We should have one for sysbuild @tejlmand

File: cmake/modules/extensions.cmake [hunk 1]
```diff
@@ -2594,6 +2594,10 @@ Relative paths are only allowed with `-D${ARGV1}=<path>`")
         set(ZFILE_BOARD_IDENTIFIER ${BOARD_IDENTIFIER})
       endif()
     endif()
+    if(NOT DEFINED ZFILE_SUFFIX)
+      # Defaulting to system wide settings when SUFFIX is not given as argument
+      set(ZFILE_SUFFIX ${FILE_SUFFIX})
+    endif()
```
Comment (tejlmand) [inline cmake/modules/extensions.cmake:2600]: this change impact behavior for cases where suffix should not be applied, such as board and shield dir files.
https://github.com/zephyrproject-rtos/zephyr/blob/9a0b8e7809b0e877d2be664dcc09f9ba117b8a19/cmake/modules/dts.cmake#L157
https://github.com/zephyrproject-rtos/zephyr/blob/9a0b8e7809b0e877d2be664dcc09f9ba117b8a19/cmake/modules/shields.cmake#L118
https://github.com/zephyrproject-rtos/zephyr/blob/9a0b8e7809b0e877d2be664dcc09f9ba117b8a19/cmake/modules/shields.cmake#L122

**Review by tejlmand — CHANGES_REQUESTED** [review]
we need to decide on how to use `FILE_SUFFIX` per default.

As now, it's an application feature so that application may have different configuration depending on use-case.

File suffix'es should not be applied to board / shield dirs under board_root, as those has variants and revisions.

Thus defaulting to pickup suffix'es is not ideal.

File: cmake/modules/extensions.cmake [hunk 2]
```diff
@@ -2535,75 +2535,75 @@ Please provide one of following: APPLICATION_ROOT, CONF_FILES")
     set(multi_args CONF_FILES NAMES)
   endif()
 
-  cmake_parse_arguments(FILE "${options}" "${single_args}" "${multi_args}" ${ARGN})
-  if(FILE_UNPARSED_ARGUMENTS)
-      message(FATAL_ERROR "zephyr_file(${ARGV0} <val> ...) given unknown arguments: ${FILE_UNPARSED_ARGUMENTS}")
+  cmake_parse_arguments(ZFILE "${options}" "${single_args}" "${multi_args}" ${ARGN})
```
[the remark this answers is not in the document]
Comment (57300) [reply cmake/modules/extensions.cmake:2538]: > When I first looked at the commit it wasn't clear to me that this was actually preparation work to allow picking up `FILE_SUFFIX` from parent scope.

Not exactly. As a standalone change, it has additional, unrelated effects:
* Prevent `FILE_SUFFIX` from being (incorrectly) inherited from cache scope
* Support `zephyr_boilerplate_watch(FILE_SUFFIX)` if we ever decide we need it

I'll try to rephrase the commit message though.

File: cmake/modules/extensions.cmake (hunk 1 above)
Comment (57300) [reply cmake/modules/extensions.cmake:2600]: This behavior was already present due to `FILE_SUFFIX` being inherited from cache scope.

If we don't want to apply the suffix in these places, then I can fix it now by dropping one of the commits and instead adding explicit `SUFFIX ${FILE_SUFFIX}` where it's needed for sure.

File: cmake/modules/extensions.cmake (hunk 2 above)
[the remark this answers is not in the document]
Comment (57300) [reply cmake/modules/extensions.cmake:2538]: > > then any parent or cached scoped FILE_SUFFIX is no longer available, even when SUFFIX is not passed as argument to zephyr_file().
> 
> Oh my, it does seem like cache variables can slip through. Guess a CMake issue should be raised.

Is that unexpected? I figured that if `SUFFIX` was not passed, then `cmake_parse_arguments()` would set `FILE_SUFFIX` to undefined in local scope, which is why the cached value slipped through.

[the remark this answers is not in the document]
Comment (tejlmand) [reply cmake/modules/extensions.cmake:2538]: https://gitlab.kitware.com/cmake/cmake/-/issues/25773

File: cmake/modules/extensions.cmake (hunk 1 above)
Comment (tejlmand) [reply cmake/modules/extensions.cmake:2600]: I think we should avoid having FILE_SUFFIX applied on boards as that will open for misuse.
Instead hwmv2 provides board variant options.

Comment (nordicjm) [reply cmake/modules/extensions.cmake:2600]: Agreed, files in `<board_dir>/<vendor>/<board>` should not have this applied, it should be applied to `<app>/<boards>/<board>.overlay` and `<app>/<boards>/<board>.dts` files

File: cmake/modules/configuration_files.cmake [hunk 3]
```diff
@@ -73,9 +73,10 @@ To change CONF_FILE, use the CONF_FILE variable." ${CONF_FILE_FORCE_CACHE})
 # The CONF_FILE variable is now set to its final value.
 zephyr_boilerplate_watch(CONF_FILE)
 
-zephyr_file(CONF_FILES ${APPLICATION_CONFIG_DIR}/boards DTS APP_BOARD_DTS SUFFIX ${FILE_SUFFIX})
-
 zephyr_get(DTC_OVERLAY_FILE SYSBUILD LOCAL)
+if(NOT DEFINED DTC_OVERLAY_FILE)
+  zephyr_file(CONF_FILES ${APPLICATION_CONFIG_DIR}/boards DTS DTC_OVERLAY_FILE SUFFIX ${FILE_SUFFIX})
+endif()
```
[the remark this answers is not in the document]
Comment (57300) [reply cmake/modules/configuration_files.cmake:82]: Added comments.

File: cmake/modules/extensions.cmake (hunk 1 above)
Comment (57300) [reply cmake/modules/extensions.cmake]: Done.

**Review by nordicjm — COMMENTED** [review]
Problem I have with this is it seems impossible to use a file suffix with just sysbuild e.g. `-Dsysbuild_FILE_SUFFIX=<x>`

File: share/sysbuild/cmake/modules/sysbuild_kconfig.cmake [hunk 4]
```diff
@@ -17,11 +17,9 @@ if(DEFINED SB_CONF_FILE)
   # SB_CONF_FILE already set so nothing to do.
 elseif(DEFINED ENV{SB_CONF_FILE})
   set(SB_CONF_FILE $ENV{SB_CONF_FILE})
-elseif(EXISTS      ${APP_DIR}/sysbuild.conf)
-  set(SB_CONF_FILE ${APP_DIR}/sysbuild.conf)
 else()
-  # Because SYSBuild is opt-in feature, then it is permitted to not have a
-  # SYSBuild dedicated configuration file.
+  # sysbuild.conf is an optional file, because sysbuild is an opt-in feature.
+  zephyr_file(CONF_FILES ${APP_DIR} KCONF SB_CONF_FILE NAMES "sysbuild.conf")
```
Comment (nordicjm) [inline share/sysbuild/cmake/modules/sysbuild_kconfig.cmake:22]: Shouldn't this have `SUFFIX` as part of it?

**Comment by 57300** [comment]
> Problem I have with this is it seems impossible to use a file suffix with just sysbuild e.g. `-Dsysbuild_FILE_SUFFIX=<x>`

It's not a bug and I don't know why we need it.

File: share/sysbuild/cmake/modules/sysbuild_kconfig.cmake (hunk 4 above)
Comment (57300) [reply share/sysbuild/cmake/modules/sysbuild_kconfig.cmake]: Yes, thanks for spotting

**Review by nordicjm — APPROVED** [review]

File: cmake/modules/extensions.cmake (hunk 2 above)
[the remark this answers is not in the document]
Comment (tejlmand) [reply cmake/modules/extensions.cmake:2538]: `ZFILE` is fine for now, but seems CMake might be opting for having a recommendation to use something like `ARG` / `arg` (not sure i'm fond of that, but let's wait and see)

Ref: https://gitlab.kitware.com/cmake/cmake/-/issues/25773#note_1499098

**Review by tejlmand — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Pull request merged**